tensorizer/serialization.py

```diff
@@ -1681,30 +1684,86 @@ def __init__(
                 structure.filter(self._metadata.__contains__)
             self._structure: Dict[Union[str, int], Any] = structure.dict()
 
-            if not isinstance(num_readers, int):
+            dynamic_num_readers: bool = num_readers is None
+            if not dynamic_num_readers and not isinstance(num_readers, int):
                 raise TypeError(
-                    "num_readers: expected int,"
+                    "num_readers: expected int or None,"
                     f" got {num_readers.__class__.__name__}"
                 )
-            elif num_readers < 1:
+
+            if dynamic_num_readers:
+                if isinstance(
+                    getattr(self._file, "raw", self._file), io.FileIO
+                ):
+                    num_readers = 8
+                elif self._verify_hash:
+                    num_readers = 4
+                else:
+                    num_readers = 2
+                if is_cuda:
+                    free_ram = psutil.virtual_memory().available
+                    allowed_ram = free_ram - (10 << 20)
+                    tensor_sizes = sorted(
+                        (
+                            t.deserialized_length
+                            for t in self._metadata.values()
+                        ),
+                        reverse=True,
+                    )
+                    num_readers = min(num_readers, len(tensor_sizes)) or 1
+                    while (
+                        num_readers > 1
+                        and sum(tensor_sizes[:num_readers]) > allowed_ram
+                    ):
+                        num_readers -= 1
+                    del tensor_sizes
```

Comment:
nit; pull out into a separate method

Reply:
Done in 452e99e842cc46261a787fcae7d2c9360320d1ad